pyccel/codegen/printing/ccode.py

```diff
@@ -666,31 +666,53 @@ def _print_IndexedElement(self, expr):
-                                isinstance(start.args[0], LiteralInteger):
-                            start = PyccelMinus(base_shape[i], start.args[0])
-                        end = ind.end
-                        if isinstance(end, PyccelUnarySub) and \
-                                isinstance(end.args[0], LiteralInteger):
-                            end = PyccelMinus(base_shape[i], end.args[0])
-                        if ind.start is None:
-                            start = 0
-                        if ind.end is None:
-                            end = base.shape[i]
-                        inds[i] = Slice(start, end)
-                    else:
-                        #setting the Slice start and end to their correct value when try to get a view with scalar index
-                        inds[i] = Slice(ind, PyccelAdd(ind, LiteralInteger(1)))
+                        inds[i] = self._slice_elements_managements(ind, PyccelArraySize(base, i),
+                            allow_negative_indexes)
                 inds = [self._print(i) for i in inds]
                 return "array_slicing(%s, %s)" % (base_name, ", ".join(inds))
             inds = [self._print(i) for i in inds]
         else:
             raise NotImplementedError(expr)
         return "%s.%s[get_index(%s, %s)]" % (base_name, dtype, base_name, ", ".join(inds))
 
+    def _slice_elements_managements(self, _slice, shape, allow_negative_index):
+
+        start = LiteralInteger(0) if _slice.start is None else _slice.start
+        end = shape if _slice.end is None else _slice.end
+
+        # negative start and end in slice
+        if isinstance(start, PyccelUnarySub) and isinstance(start.args[0], LiteralInteger):
+            start = PyccelMinus(shape, start.args[0])
+        elif allow_negative_index and not isinstance(start, LiteralInteger):
+            start = IfTernaryOperator(PyccelLt(start, LiteralInteger(0)),
+                            PyccelMinus(shape, start), start)
+
+        if isinstance(end, PyccelUnarySub) and isinstance(end.args[0], LiteralInteger):
+            end = PyccelMinus(shape, end.args[0])
+        elif allow_negative_index and not isinstance(end, LiteralInteger):
+            end = IfTernaryOperator(PyccelLt(end, LiteralInteger(0)),
+                            PyccelMinus(shape, end), end)
+
+        # steps in slices
+        step = _slice.step
+
+        # negative step in slice
+        if isinstance(step, PyccelUnarySub) and isinstance(step.args[0], LiteralInteger):
+            start = shape if _slice.start is None else start
+            end = LiteralInteger(0) if _slice.end is None else end
+
+        # variable step in slice
+        elif allow_negative_index and step and not isinstance(step, LiteralInteger):
+            start = IfTernaryOperator(PyccelGt(step, LiteralInteger(0)), start, end)
+            end = IfTernaryOperator(PyccelGt(step, LiteralInteger(0)), end, start)
+        else :
+            step = LiteralInteger(1)
```

Comment:
Am I missing something or does this miss the case of a non-negative step (e.g. `2:4:2`, or `:5:i` without `@allow_negative_index`)?

Reply:
In C as @saniQQ  said we will always check the step value if it's negative or positive either there is a decorator or non.

Reply:
That is not my point. Here you have 3 cases:
1. **Step is a negative literal**
    Here you keep the step
2. **Step is provided and the `@allow_negative_index` decorator is set**
    Here you keep the step
3. **Everything else** (e.g. `step=2`)
    Here you set the step to 1

Reply:
I suspect your `else` should be: `elif not step:`

Reply:
Compare this to the original code here:
https://github.com/pyccel/pyccel/pull/588#discussion_r542381620

Reply:
... that's true, I changed the condition and I didn't pay attention here it should be 
```
elif step is None : 
       step = 1
```
or like you have mentioned above.